### DISCUSSION TITLE
Start using attrs instead of namedtuple

### DISCUSSION THREAD

**Discussion by balloob** [home-assistant/architecture#896] [General]
[attrs](http://www.attrs.org/en/stable/) is a Python package that allows you to write simple classes without any boilerplate. It's like a namedtuple, but better. 

It makes it a lot easier to write small classes to pass data around. We use namedtuple 41 times currently in our codebase.

Normally I am not the person to propose adding dependencies to core, however I noticed that aiohttp started using attrs 2 weeks ago (https://github.com/aio-libs/aiohttp/pull/2690) and thus it is already installed for a base install of Home Assistant.

Python 3.7 includes something similar to attrs called "Data Classes". However, it depends on [PEP 526](https://www.python.org/dev/peps/pep-0526/) which was only included in Python 3.6.

I think that using attrs can make our code easier to use. It also does not add any runtime costs since it does the work when the class is declared, not when a property is requested.

**Comment by balloob** [comment]
One of the advantages of attrs over namedtuple is default values. Now I tend to convert a namedtuple to a class whenever I want to add default values to avoid this: 

```python
Message = namedtuple('Message', ['topic', 'payload', 'qos', 'retain'])
Message.__new__.__defaults__ = (0, False)
```

**Comment by pvizeli** [comment]
Looks amazing. I don't like namedtuple anyway :)

**Comment by balloob** [comment]
We're using attrs now!

**Comment by frenck** [comment]
Migrated this one as a test to see if closed issues are properly migrated (including they closing state). They are not. 

Please ignore this action/comment.
